### PR TITLE
Support top-level tensor values in dataset

### LINKED ISSUE #17128: [dataset] Support tensor type blocks
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*

Dataset doesn't seem to be pytorch compatible, and treats `torch.tensor` as a generic object. This is both unintuitive and makes it difficult to use zero-copy dese. 

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

```
import ray
import torch

ray.init()
ds = ray.experimental.data.from_items([torch.tensor([0, 0]) for _ in range(400)])
first = next(ds.iter_batches(batch_format="pyarrow"))
```

Using the pandas format for the batch results in a 1 column df of type "object" (but at least doesn't crash). 

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by ericl**
This PR adds basic support for top-level Tensors, which are treated separately as pyarrow.Tensor (np.ndarray) format. Nested tensor field support is punted for future work here: https://github.com/ray-project/ray/issues/17457

TODOs:
- [x] tensor block
- [x] add read/write numpy support
- [x] testing
- [x] update docs

Closes https://github.com/ray-project/ray/issues/17128

### CODE REVIEW

**Comment by ericl** [comment]
@clarkzinzow I'm hacking on this in spare time, feel free to push improvements

File: python/ray/experimental/data/datasource/datasource.py [hunk 1]
```diff
@@ -135,28 +135,39 @@ class RangeDatasource(Datasource[Union[ArrowRow, int]]):
...
 
         # Example of a read task. In a real datasource, this would pull data
         # from an external system instead of generating dummy data.
         def make_block(start: int, count: int) -> Block:
-            if use_arrow:
+            if block_format == "arrow":
                 return pyarrow.Table.from_arrays(
                     [np.arange(start, start + count)], names=["value"])
+            elif block_format == "tensor":
+                return np.ones(tensor_shape) * np.expand_dims(
```
Comment (wuisawesome) [inline python/ray/experimental/data/datasource/datasource.py:153]: `return np.arange(start, start+count, shape=tensor_shape`?

Comment (wuisawesome) [reply python/ray/experimental/data/datasource/datasource.py:153]: I think we should be able to avoid matrix multiplication here.

**Review by wuisawesome — CHANGES_REQUESTED** [review]
Can you add a test that writes the tensor using `np.savefile` and then loads it using `ds.read_numpy()` (this type of test case wasn't actually trivial in the case of other data sources)?

**Review by ericl — COMMENTED** [review]
Test added

**Review by wuisawesome — APPROVED** [review]

**Pull request merged**